gst/nnstreamer/elements/gsttensor_trainer.c

```diff
@@ -627,126 +620,317 @@ gst_tensor_trainer_epochs_is_complete (GstTensorTrainer * trainer)
+
+  return TRUE;
+}
+
+/**
+ * @brief Append output tensors(out_mem) to outbuf.
+ */
+static void
+gst_tensor_trainer_append_output_tensor_to_outbuf (GstBuffer * outbuf,
+    GstMemory ** out_mem, guint num_tensors)
+{
+  guint i;
+  for (i = 0; i < num_tensors; i++) {
+    /* append the memory block(output_tensors) to outbuf */
+    gst_buffer_append_memory (outbuf, out_mem[i]);
+  }
+  GST_INFO ("out_buffer size : %zd", gst_buffer_get_size (outbuf));
+}
+
+/**
+ * @brief Create output tensors.
+ */
+static gboolean
+gst_tensor_trainer_create_output_tensors (GstTensorTrainer * trainer,
+    GstMemory ** out_mem, GstMapInfo * out_info)
+{
+  guint i;
+  void *ptr;
+  GstTensorMemory out_tensors[NNS_TENSOR_SIZE_LIMIT];
+  double model_stats[MODEL_STATS_SIZE] =
+      { -INFINITY, -INFINITY, -INFINITY, -INFINITY };
+
+  for (i = 0; i < trainer->output_meta.num_tensors; i++) {
+    out_tensors[i].data = NULL;
+    out_tensors[i].size =
+        gst_tensor_trainer_get_tensor_size (trainer, i, FALSE);
+
+    out_mem[i] = gst_allocator_alloc (NULL, out_tensors[i].size, NULL);
+    if (!out_mem[i]) {
+      GST_ERROR_OBJECT (trainer, "Failed to allocate memory");
+      return FALSE;
+    }
+
+    if (!gst_memory_map (out_mem[i], (out_info + i), GST_MAP_WRITE)) {
+      GST_ERROR_OBJECT (trainer, "Could not map in_mem[%u] GstMemory", i);
+      return FALSE;
+    }
+
+    out_tensors[i].data = out_info[i].data;
+
+    if (!gst_tensor_trainer_get_model_stats (trainer, model_stats))
+      return FALSE;
+
+    ptr = out_tensors[i].data;
+    memcpy (ptr, model_stats, sizeof (model_stats));
+  }
+
+  /* Unmap out_info */
+  for (i = 0; i < trainer->output_meta.num_tensors; i++) {
+    if (out_mem[i])
+      gst_memory_unmap (out_mem[i], (out_info + i));
+  }
+
+  return TRUE;
+}
+
+/**
+ * @brief Chain function, this function does the actual processing.
+ */
+static GstFlowReturn
+gst_tensor_trainer_chain (GstPad * sinkpad, GstObject * parent,
+    GstBuffer * inbuf)
+{
+  GstTensorTrainer *trainer;
+  GstBuffer *outbuf = NULL;
+  gint ret = -1;
+  guint num_tensors;
+  gboolean in_flexible;
+  GstMemory *in_mem[NNS_TENSOR_SIZE_LIMIT] = { 0, };
+  GstMapInfo in_info[NNS_TENSOR_SIZE_LIMIT];
+  GstMemory *out_mem[NNS_TENSOR_SIZE_LIMIT] = { 0, };
+  GstMapInfo out_info[NNS_TENSOR_SIZE_LIMIT];
+
+  trainer = GST_TENSOR_TRAINER (parent);
+  in_flexible = gst_tensor_pad_caps_is_flexible (sinkpad);
+  num_tensors = gst_tensor_buffer_get_count (inbuf);
+
+  if (!gst_tensor_trainer_check_chain_conditions (trainer, num_tensors))
+    return GST_FLOW_ERROR;
+
+  if (gst_tensor_trainer_check_buffer_drop_conditions (trainer))
+    return GST_FLOW_OK;
+
+  if (!gst_tensor_trainer_create_input_tensors (trainer, inbuf,
+          (GstMemory **) & in_mem, in_info, in_flexible))
+    goto error;
+
+  if (FALSE == gst_tensor_trainer_check_input_tensors_size (trainer,
+          trainer->input_tensors, num_tensors, in_flexible))
```

Comment:
Is `trainer->input_tensors` necessary?

Reply:
Since the API name is to check input tensor size, I entered input_tensors as input.
Should we handle GstTensorsMemory internally in the API?
```
static gboolean
gst_tensor_trainer_check_input_tensors_size (GstTensorTrainer * trainer,
    guint num_tensors, gboolean in_flexible)
{
  guint32 i;
  gsize expected;
 GstTensorMemory * tensors = trainer->input_tensors;
 }
 ```

Reply:
IMO, It looks better.